pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -247,6 +259,43 @@ func (r *reconciler) reconcile(ctx context.Context, req reconcile.Request, log *
 	return nil
 }
 
+func finalizeIfNeeded(ctx context.Context, stream *imagev1.ImageStream, srcClusterName string, clients map[string]ctrlruntimeclient.Client) error {
+	finalizerSet := sets.NewString(stream.ObjectMeta.Finalizers...)
+	if finalizerSet.Has(ControllerName) {
+		for clusterName, client := range clients {
+			if clusterName == srcClusterName {
+				continue
+			}
+			// populate deleting to all clusters except srcClusterName
+			if err := client.Delete(ctx, &imagev1.ImageStream{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      stream.Name,
+					Namespace: stream.Namespace,
+				},
+			}); !apierrors.IsNotFound(err) {
+				return err
+			}
+		}
+
+		// remove our finalizer from the list and update it.
+		stream.ObjectMeta.Finalizers = finalizerSet.Difference(sets.NewString(ControllerName)).List()
+		if err := clients[srcClusterName].Update(ctx, stream); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func ensureFinalizer(ctx context.Context, stream *imagev1.ImageStream, client ctrlruntimeclient.Client) error {
+	if !sets.NewString(stream.ObjectMeta.Finalizers...).Has(ControllerName) {
```

Comment:
stylistic nit: Prefer left-aligned code, so here that would be:
```
if sets.NewString(stream.Finalizers).Has(ControllerName) {
  return nil
}
stream.Finalizers = append(stream.Finalizers, ControllerName)
return client.Update(ctx, stream)
```
That both saves you some lines and makes the code quicker to read.

See https://blog.codinghorror.com/flattening-arrow-code/ for some more examples of this

Reply:
Updated.